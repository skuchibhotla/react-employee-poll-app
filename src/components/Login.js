import {connect} from "react-redux";
import {Navigate} from "react-router-dom";
import {useState} from "react";
import {handleLogin} from "../actions/authedUser";

const Login = ({dispatch, loggedIn}) => {
    // Starting with blank state for username and password...
    const [username, setUsername] = useState("");
    const [password, setPassword] = useState("");
    const [errorMessage, setErrorMessage] = useState("");

    if (loggedIn) {
        const urlParams = new URLSearchParams(window.location.search);
        const redirectUrl = urlParams.get('redirectTo');
        return <Navigate to={redirectUrl ? redirectUrl : "/"}/>;
    }

    const handleUsername = (e) => {
        const value = e.target.value;
        setUsername(value);
    };

    const handlePassword = (e) => {
        const value = e.target.value;
        setPassword(value);
    };

    const handleSubmit = (e) => {
        e.preventDefault();

        if (!username || !password) {
            setErrorMessage("Error: Username and password are required!"); // Set error message...
            return; // Stop the function execution...
        }

        dispatch(handleLogin(username, password));
        setUsername("");
        setPassword("");
    };

    return (
        <div data-testid="custom-login">
            <h1 className="text-3xl font-bold mt-9" data-testid="login-heading">Login</h1>
            
            {/* Error message... */}
            {errorMessage && <div className="error" data-testid="error-message">{errorMessage}</div>}


            <form onSubmit={handleSubmit}>
                {/* User Name */}
                <div className="form-group">
                    <label htmlFor="username">Username</label>
                        <input className="form-control"
                            value={username}
                            onChange={handleUsername}
                            type="text"
                            name="username"
                            id="username"
                            data-testid="username"
                        />
                </div>

                {/* Password */}
                <div className="form-group">
                    <label htmlFor="password">Password</label>
                        <input className="form-control"
                            value={password}
                            onChange={handlePassword}
                            type="password"
                            name="password"
                            id="password"
                            data-testid="password"
                        />
                </div>
                {/* Submit Button */}
                <button type="submit" className="btn btn-primary" data-testid="submit">Login</button>
            </form>
        </div>
    );
};

const mapStateToProps = ({authedUser}) => ({
    loggedIn: !!authedUser,
});

export default connect(mapStateToProps)(Login);
