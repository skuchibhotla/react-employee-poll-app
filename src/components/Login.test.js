import {fireEvent, render} from "@testing-library/react";
import {Provider} from "react-redux";
import {store} from "../store";
import {BrowserRouter} from "react-router-dom";
import React from "react";
import Login from "./Login";
import {handleInitialData} from "../actions/shared";

describe("Login", () => {
    it("should render the component", () => {
        const component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Login/>
                </BrowserRouter>
            </Provider>
        );
        expect(component).toBeDefined();
        expect(component).toMatchSnapshot();
    });

    it('should update input fields when user types and validate their presence', async () => {
        await store.dispatch(handleInitialData());

        const wrapper = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Login/>
                </BrowserRouter>
            </Provider>
        );

        // Define username, password, and submit button...
        const usernameInputElement = wrapper.getByTestId("username");
        const passwordInputElement = wrapper.getByTestId("password");
        const submitButtonElement = wrapper.getByTestId("submit");

        // Expects...
        expect(usernameInputElement).toBeInTheDocument();
        expect(passwordInputElement).toBeInTheDocument();
        expect(submitButtonElement).toBeInTheDocument();

        fireEvent.change(usernameInputElement, {target: {value: 'tylermcginnis'}});
        fireEvent.change(passwordInputElement, {target: {value: 'test'}});
        
        expect(usernameInputElement.value).toBe("tylermcginnis");
        expect(passwordInputElement.value).toBe("test");
    });

    it('should show an error message if submit with empty fields', async () => {
        // Your setup code for rendering the component...

        const wrapper = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Login/>
                </BrowserRouter>
            </Provider>
        );

        // Get submit button and click it without filling the fields
        const submitButtonElement = wrapper.getByTestId("submit");
        fireEvent.click(submitButtonElement);

        // Check if the error message is displayed
        const errorMessage = wrapper.queryByTestId("error-message");
        expect(errorMessage).toBeInTheDocument();
        expect(errorMessage.textContent).toBe("Error: Username and password are required!");
    });
});