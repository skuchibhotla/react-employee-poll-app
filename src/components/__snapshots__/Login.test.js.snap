// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should render the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        data-testid="custom-login"
      >
        <h1
          class="text-3xl font-bold mt-9"
          data-testid="login-heading"
        >
          Login
        </h1>
        <form>
          <div
            class="form-group"
          >
            <label
              for="username"
            >
              Username
            </label>
            <input
              class="form-control"
              data-testid="username"
              id="username"
              name="username"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="password"
            >
              Password
            </label>
            <input
              class="form-control"
              data-testid="password"
              id="password"
              name="password"
              type="password"
              value=""
            />
          </div>
          <button
            class="btn btn-primary"
            data-testid="submit"
            type="submit"
          >
            Login
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      data-testid="custom-login"
    >
      <h1
        class="text-3xl font-bold mt-9"
        data-testid="login-heading"
      >
        Login
      </h1>
      <form>
        <div
          class="form-group"
        >
          <label
            for="username"
          >
            Username
          </label>
          <input
            class="form-control"
            data-testid="username"
            id="username"
            name="username"
            type="text"
            value=""
          />
        </div>
        <div
          class="form-group"
        >
          <label
            for="password"
          >
            Password
          </label>
          <input
            class="form-control"
            data-testid="password"
            id="password"
            name="password"
            type="password"
            value=""
          />
        </div>
        <button
          class="btn btn-primary"
          data-testid="submit"
          type="submit"
        >
          Login
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
